File: Data Structures & Algorithms/search-2d-matrix/submission-2.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        int n = matrix.Length, m = matrix[0].Length;
        int low = 0, high = n*m-1;

        while (low <= high){
            int middle = low + (high-low) / 2;
            int value = matrix[middle / m][middle % m];
            
            if (value == target) return true;
            if (value < target) low = middle+1;
            else high = middle-1;
        }

        return false;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-2.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1, high = *max_element(piles.begin(), piles.end());

        while (low <= high){     
            int mid = low + (high-low) / 2;
            int time = 0;
            for (int pile : piles){
                time += ceil((double) pile / mid);
            }

            if (time <= h) high = mid - 1;
            else low = mid + 1; 
        }

        return low;
    }
};
